Adding microbenchmarks for encoder/decoder.

nmittler opened the pull request:
Fixes #23

nmittler commented:
@jpinner I went with the multi-module approach ... just thought it was simpler than profiles with a custom artifact.

This code was taken from the work @Scottmitch had done in the [Netty Http2FrameWriterBenchmark](https://github.com/netty/netty/blob/master/microbench/src/main/java/io/netty/microbench/http2/Http2FrameWriterBenchmark.java).

/cc @Scottmitch @louiscryan

nmittler commented:
Early results on my machine:

```
*** ENCODER ***
Benchmark                (maxTableSize)  (sensitive)  (size)   Mode  Cnt      Score     Error  Units
EncoderBenchmark.encode            4096         true   SMALL  thrpt  100   8964.311 ± 159.581  ops/s
EncoderBenchmark.encode            4096         true  MEDIUM  thrpt  100   9039.056 ± 230.622  ops/s
EncoderBenchmark.encode            4096         true   LARGE  thrpt  100   7649.271 ± 172.577  ops/s
EncoderBenchmark.encode            4096         true   JUMBO  thrpt  100   2372.141 ±  49.844  ops/s
EncoderBenchmark.encode            4096        false   SMALL  thrpt  100   9170.711 ± 153.452  ops/s
EncoderBenchmark.encode            4096        false  MEDIUM  thrpt  100  10642.594 ± 288.204  ops/s
EncoderBenchmark.encode            4096        false   LARGE  thrpt  100   5915.217 ±  53.251  ops/s
EncoderBenchmark.encode            4096        false   JUMBO  thrpt  100   1148.010 ±  12.330  ops/s

*** DECODER ***
Benchmark                (maxHeaderSize)  (maxTableSize)  (sensitive)  (size)   Mode  Cnt        Score       Error  Units
DecoderBenchmark.decode             8192            4096         true   SMALL  thrpt  100  1759325.052 ± 25195.805  ops/s
DecoderBenchmark.decode             8192            4096         true  MEDIUM  thrpt  100   518582.163 ±  2787.239  ops/s
DecoderBenchmark.decode             8192            4096         true   LARGE  thrpt  100   101328.476 ±   538.392  ops/s
DecoderBenchmark.decode             8192            4096         true   JUMBO  thrpt  100    18312.158 ±   204.007  ops/s
DecoderBenchmark.decode             8192            4096        false   SMALL  thrpt  100  1678156.810 ±  9052.228  ops/s
DecoderBenchmark.decode             8192            4096        false  MEDIUM  thrpt  100   488803.678 ±  1789.771  ops/s
DecoderBenchmark.decode             8192            4096        false   LARGE  thrpt  100    88952.083 ±  1173.748  ops/s
DecoderBenchmark.decode             8192            4096        false   JUMBO  thrpt  100    16978.573 ±   283.292  ops/s
```

Scottmitch commented on microbench/main/java/com.twitter.hpack.microbench/DecoderBenchmark.java:
```diff
@@ -0,0 +1,96 @@
+        decoder.decode(new ByteArrayInputStream(input), new HeaderListener() {
+            @Override
+            public void addHeader(byte[] name, byte[] value, boolean sensitive) {
+                // Do nothing.
```
I wonder if this will get optimized away?  Perhaps we should do something in here...`BlackHole` or something?

nmittler replied:
Done.

Scottmitch commented:
@nmittler - Not sure if you saw https://github.com/netty/netty/issues/3503#issuecomment-87782071 but that describes some stuff that may be limiting with what I had in Netty.  Looks like you already are exercising the `sensitive` parameter but some of the other items may still be relevant.

nmittler commented:
@Scottmitch good point.  In particular your points 1 and 2 might be worth exploring in these benchmarks.

@buchgr in your benchmarking for grpc have you put any thought into "representational" header name/values?

buchgr commented:
@nmittler so you are asking for example headers? I collected some grpc headers for my headers benchmarks https://github.com/netty/netty/pull/3681/files#diff-ca4f23f3c28ec164487afb498dc26e5dR37 other than that I don't.

buchgr commented:
@nmittler at least in GRPC we are not spending much time in hpack in our load tests (~2%), but then again we also don't use many headers.
![hpack](https://cloud.githubusercontent.com/assets/1388179/7330103/742f7d02-ea9d-11e4-9311-ddb13a945d4a.png)
The decode and encode methods are definitely a good starting point for benchmarking.

Before starting to do micro optimizations it would likely make sense to do some load tests with a small client / server that uses lots of headers and is build on top of Netty's HTTP/2 codec. You could then look at it with a profiler and see what the total optimization potential is e.g. you could make hpack encode twice as fast, but if we only every spend 2% of our time in hpack ...

tatsuhiro-t commented:
Since Encoder only uses huffman if it makes string strictly shorter, Encoder does not use huffman encoding in Decoder benchmark since we have 0-255 value range for each byte and not frequently used byte not in ascii range is very long huffman code.  To get more realistic benchmark, perhaps we can limit the random value range into printable ascii?

jpinner commented:
When I did some benchmarking I used two sets of random strings -- one across all range of bytes and one limited to ascii characters.

nmittler commented:
PTAL, I've added an extra parameter to `limitToAscii` where I use a reduced ASCII character set (upper/lowercase and a couple of other characters). I've updated the results with these changes.

nmittler commented:
@buchgr I understand that grpc isn't spending a lot of time in hpack, but I still think it's useful to have the benchmarks to get a basic feel for how well it performs overall and so that we have a basis for comparison going forward.

buchgr commented:
@nmittler yeah absolutely. I think those benchmarks are great. I guess what I wanted to say is that, before as a next step you and @Scottmitch start spending time on micro optimizations it might make sense to see what the optimization potential is, in order to see if this is high priority or not. My guess would be that, since this stuff runs in production at Twitter it likely performs pretty decent.

Scottmitch commented:
@buchgr - +1.

@jpinner - Are you using this library with your HTTP/2 codec that you open-sourced a while back, or some thing else?  I'm curious what your interface looks like to this (do you have `InputStream`, `OutputStream`, and `byte[]` handy, or is there some amount of copy going on)?

jpinner commented:
@Scottmitch we use this with the http/2 codec. For decoding there is some copying on decode into a "cumulation" buffer like in most of the netty frame decoders and then we wrap the buffer in a ByteBufInputStream. For encoding we create a ByteBufOutputStream. See:

https://github.com/twitter/netty-http2/blob/master/src/main/java/com/twitter/http2/HttpHeaderBlockDecoder.java

https://github.com/twitter/netty-http2/blob/master/src/main/java/com/twitter/http2/HttpHeaderBlockEncoder.java

nmittler commented:
@buchgr +1 :)

@jpinner thanks for the links!

@Scottmitch I've raised https://github.com/netty/netty/issues/3700 to compare Netty against the Twitter encoder/decoder.

nmittler commented:
@buchgr @Scottmitch @jpinner any other changes you'd like to see?

buchgr commented on microbench/main/java/com.twitter.hpack.microbench/DecoderBenchmark.java:
```diff
@@ -0,0 +1,82 @@
+            public void addHeader(byte[] name, byte[] value, boolean sensitive) {
+                bh.consume(name);
+                bh.consume(value);
+                bh.consume(sensitive);
```
@nmittler it should be enough to consume just one i.e. the boolean?

nmittler replied:
done.

buchgr commented on microbench/main/java/com.twitter.hpack.microbench/EncoderBenchmark.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Benchmark
+    @BenchmarkMode(Mode.Throughput)
+    public void encode() throws IOException {
```
@nmittler return the `outputStream` to make sure that computations can't be deleted by the compiler?

nmittler replied:
done.

buchgr commented:
@nmittler LGTM.

Scottmitch commented on microbench/main/java/com.twitter.hpack.microbench/AbstractMicrobenchmarkBase.java:
```diff
@@ -0,0 +1,210 @@
+
+    private static final Map<HeadersKey, List<Header>> headersMap;
+    static {
+        headersMap = new HashMap<HeadersKey, List<Header>>();
```
Initialize size to `HeadersSize.values()`?

nmittler replied:
done.

Scottmitch commented on microbench/main/java/com.twitter.hpack.microbench/EncoderBenchmark.java:
```diff
@@ -0,0 +1,68 @@
+        OutputStream outputStream = new ByteArrayOutputStream(1048576);
+        for (int i = 0; i < headers.size(); ++i) {
+            // If duplicates is set, re-add the same header each time.
+            Header header = duplicates ? headers.get(0) : headers.get(i);
```
Maybe have 2 loops based upon this condition?  We don't need the conditional overhead in each loop iteration to be part of the `benchmark`, and `duplicates` shouldn't be changing in the benchmark.

nmittler replied:
done.

Scottmitch commented:
@nmittler - LGTM!

jpinner commented on microbench/main/java/com.twitter.hpack.microbench/AbstractMicrobenchmarkBase.java:
```diff
@@ -0,0 +1,211 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.twitter.hpack.microbench;
```
is there a reason these files don't follow the maven directory hierarchy?

the netty version uses the src/test/java/pacakge-name.../microbench directory structure

nmittler replied:
@jpinner oof, good catch! I've replaced `com.twitter.hpack.microbench` with `com/twitter/hpack/microbench`.

jpinner replied:
does it also need to be moved under src/test like in netty?

On Tue, Apr 28, 2015 at 1:26 PM, Nathan Mittler <notifications@github.com>
wrote:

> In
> microbench/main/java/com.twitter.hpack.microbench/AbstractMicrobenchmarkBase.java
> <https://github.com/twitter/hpack/pull/24#discussion_r29284441>:
>
> > +/*
> > + * Copyright 2015 Twitter, Inc.
> > + *
> > + * Licensed under the Apache License, Version 2.0 (the "License");
> > + * you may not use this file except in compliance with the License.
> > + * You may obtain a copy of the License at
> > + *
> > + *     http://www.apache.org/licenses/LICENSE-2.0
> > + *
> > + * Unless required by applicable law or agreed to in writing, software
> > + * distributed under the License is distributed on an "AS IS" BASIS,
> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > + * See the License for the specific language governing permissions and
> > + * limitations under the License.
> > + */
> > +package com.twitter.hpack.microbench;
>
> @jpinner <https://github.com/jpinner> oof, good catch! I've replaced
> com.twitter.hpack.microbench with com/twitter/hpack/microbench.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/hpack/pull/24/files#r29284441>.
>

nmittler replied:
Actually, Netty has it under `src/main`: https://github.com/netty/netty/tree/master/microbench/src

jpinner replied:
so we need to move it under "src" still?

On Tue, Apr 28, 2015 at 1:38 PM, Nathan Mittler <notifications@github.com>
wrote:

> In
> microbench/main/java/com.twitter.hpack.microbench/AbstractMicrobenchmarkBase.java
> <https://github.com/twitter/hpack/pull/24#discussion_r29285752>:
>
> > +/*
> > + * Copyright 2015 Twitter, Inc.
> > + *
> > + * Licensed under the Apache License, Version 2.0 (the "License");
> > + * you may not use this file except in compliance with the License.
> > + * You may obtain a copy of the License at
> > + *
> > + *     http://www.apache.org/licenses/LICENSE-2.0
> > + *
> > + * Unless required by applicable law or agreed to in writing, software
> > + * distributed under the License is distributed on an "AS IS" BASIS,
> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > + * See the License for the specific language governing permissions and
> > + * limitations under the License.
> > + */
> > +package com.twitter.hpack.microbench;
>
> Actually, Netty has it under src/main:
> https://github.com/netty/netty/tree/master/microbench/src
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/hpack/pull/24/files#r29285752>.
>

nmittler replied:
Ah missed that ... done.

Scottmitch commented on microbench/src/main/java/com/twitter/hpack/microbench/EncoderBenchmark.java:
```diff
@@ -0,0 +1,75 @@
+    @BenchmarkMode(Mode.Throughput)
+    public void encode(Blackhole bh) throws IOException {
+        Encoder encoder = new Encoder(maxTableSize);
+        OutputStream outputStream = new ByteArrayOutputStream(1048576);
```
@nmittler - Is this number significant, or is it just `1 << 20` which is assumed to be big enough?  Maybe add a comment?

nmittler replied:
See previous reply.

Scottmitch commented on microbench/src/main/java/com/twitter/hpack/microbench/DecoderBenchmark.java:
```diff
@@ -0,0 +1,80 @@
+
+    private static byte[] getSerializedHeaders(List<Header> headers, boolean sensitive) throws IOException {
+        Encoder encoder = new Encoder(4096);
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream(1048576);
```
This looks like the same number from https://github.com/twitter/hpack/pull/24/files#r29295744.  Any relationship, worth adding a comment or a shared `static final`?

nmittler replied:
Added construction of the output stream to the size enum.

nmittler commented:
@Scottmitch @jpinner PTAL

jpinner commented on hpack/pom.xml:
```diff
@@ -0,0 +1,29 @@
+    <parent>
+        <groupId>com.twitter</groupId>
+        <artifactId>hpack-parent</artifactId>
+        <version>0.10.2-SNAPSHOT</version>
```
let's update this to 0.11.0-SNAPSHOT

nmittler replied:
done.

jpinner commented on microbench/pom.xml:
```diff
@@ -0,0 +1,34 @@
+    <parent>
+        <groupId>com.twitter</groupId>
+        <artifactId>hpack-parent</artifactId>
+        <version>0.10.2-SNAPSHOT</version>
```
same

nmittler replied:
done.

jpinner commented on microbench/src/main/java/com/twitter/hpack/microbench/AbstractMicrobenchmarkBase.java:
```diff
@@ -0,0 +1,186 @@
+import org.openjdk.jmh.runner.options.ChainedOptionsBuilder;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+import java.io.File;
```
nit: java first (I'll try to get checkstyle running on this repo at some point)

nmittler replied:
done.

jpinner commented on microbench/src/main/java/com/twitter/hpack/microbench/HeadersSize.java:
```diff
@@ -0,0 +1,46 @@
+ */
+package com.twitter.hpack.microbench;
+
+import java.io.*;
```
nit: no globs

nmittler replied:
done.

jpinner commented on microbench/src/main/java/com/twitter/hpack/microbench/package-info.java:
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright 2014 Twitter, Inc.
```
2015

nmittler replied:
done.

jpinner commented on pom.xml:
```diff
@@ -2,10 +2,11 @@
     <groupId>com.twitter</groupId>
-    <artifactId>hpack</artifactId>
+    <artifactId>hpack-parent</artifactId>
     <version>0.10.2-SNAPSHOT</version>
```
0.11.0-SNAPSHOT

nmittler replied:
done.

jpinner commented on pom.xml:
```diff
@@ -88,6 +68,40 @@
+            <dependency>
+                <groupId>com.google.code.gson</groupId>
+                <artifactId>gson</artifactId>
+                <version>2.2.4</version>
```
while you're here: 2.3.1

nmittler replied:
done.

jpinner commented on pom.xml:
```diff
@@ -88,6 +68,40 @@
+            <dependency>
+                <groupId>junit</groupId>
+                <artifactId>junit</artifactId>
+                <version>4.11</version>
```
while you're here: 4.12

nmittler replied:
done.

jpinner commented:
@nmittler looks great! just a few nits above -- thanks for getting this all set up!

Scottmitch commented on microbench/src/main/java/com/twitter/hpack/microbench/HeadersSize.java:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Enum that indicates the size of the headers to be used for the benchmark.
+ */
+public enum HeadersSize {
```
Class is private but all methods are package protected.  Should we just make class package protected and methods public?

nmittler replied:
It has to be public for use with JMH.  I can make the methods public for consistency.

Scottmitch replied:
I see. Yah just make the methods public too I guess.

Scottmitch commented on microbench/src/main/java/com/twitter/hpack/microbench/HeadersSize.java:
```diff
@@ -0,0 +1,46 @@
+    MEDIUM(20, 40, 80),
+    LARGE(100, 100, 300);
+
+    final int numHeaders;
```
Should these be private?

nmittler replied:
done.

Scottmitch commented:
@nmittler - LGTM! Just a few cleanup questions.

nmittler commented:
@Scottmitch @jpinner committed changes ... anything else?

nmittler commented:
Great! thanks for the review, guys! :)
